Initial Envoy TLS support

jrajahalme opened the pull request:
Support TLS termination and origination in the cilium host proxy.

Current implementation embeds secrets as inline strings to (new) `K8sSecret.Name` field.

Testing is currently limited to Envoy integration testing and a Cilium bash test reaching out to an external web site.

TODO:

- [x] Fetch and maintain k8s secrets
- [ ] Add HTTP header matching support for a value derived from a `K8sSecret`
- [ ] Redact log messages containing secret values

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9263)
<!-- Reviewable:end -->

willdeuschle commented on pkg/policy/api/l4.go:
```diff
@@ -55,6 +55,36 @@ func (p PortProtocol) Covers(other PortProtocol) bool {
+	// Certificate is a chain of certificates presented to the remote party.
+	// If specified for an originating TLS context, then this is used as a
+	// client certificate
+	Certificate K8sSecret `json:"certificate,omitempty"`
```
@jrajahalme is the intention for these to be consumable only as k8s secrets? or would it also be possible to mount cert material to be used as `TLSContext`

aanm replied:
> @jrajahalme is the intention for these to be consumable only as k8s secrets?

yes

> or would it also be possible to mount cert material to be used as `TLSContext`

@willdeuschle where how/would you mount it into cilium?

willdeuschle replied:
@aanm Our configuration would be static enough such that we could deploy the agent with the certs mounted as volumes. This isn't a hard and fast requirement, it just seems like it would be lower lift.

jrajahalme committed:
tests: (WIP) TLS test

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
api: Add API support for k8s secrets

Add API definitions for k8s secrets and plumb them into Envoy policy
updates.

First implement secrets as inline strings in the `Name` field.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Update image

jrajahalme committed:
policy: Make TLS per Ep

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm committed:
pkg/k8s: add wrapper to get k8s secrets

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
add certificate manager to find certificates either locally or in k8s

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme committed:
logfields: Add tag for cached xDS version.

Add and use a tag for cached xDS version.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Always use IstioNodeToIP function

Envoy Node ID to IP mapping function is configurable, but we always
use the Istio mapping anyway. Rather than spread the configurability
elsewhere, always use the Istio convention.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
xds: Allow endpoints to wait for the current policy version to be acked

Even when no policy change is needed, the endpoint may need to wait
for the current policy version to be acked by the proxy.

Use a different Envoy node ID for proxylib, so that proxylib ACK
arriving before the main Envoy ACK does not trigger DNS response to be
released to the source POD.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Envoy: Track last ACKed version per proxy node

Track ACKs sent by proxy nodes in the xDS resource cache. This is
needed to be able to wait for the current cached resource version to
be realized by all relevant nodes.

Node entry in the map of acked versions is deleted when the network
policy for the endpoint is removed, which happens when the endpoint is
deleted.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
add 'certificates-directory' option to look for TLS certificates

This option will set the root certificate used to search for TLS
certificates defined in a CNP for L7 TLS policy enforcement.

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme committed:
envoy: Update for TLS support

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
fixes
